app/lib/core/enums/auth_result_status.dart

```diff
@@ -0,0 +1,10 @@
+enum ResultStatus {
```

Comment:
@holtzmak ... though the example you gave me an idea how to use then catch, instead of try catch, I still don't know how I can incorporate the "Optional" my current code. 

```
enum ResponseType { success, failure, catastrophic_failure }

class Response<T> { 
    final ResponseType type;
    final String reason;
    final Optional<T> response;

    Response.failure({@required this.type, @required this.reason}) :
    response = Optional.empty();

    Response.success({@required this.response}):
    type = ResponseType.success,
    reason = "Successful";
}
```

The exception is simple enough for me to understand, but how about the response type? Using what I have right now, can you give an example how to do this? Let's use the signIn method as an example

 ```
Future<ResultStatus> signIn({
    @required String email,
    @required String password,
  }) async {
    try {
      UserCredential authResult =
          await _firebaseAuth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      var user = await _firestoreService.fetchUser(authResult.user.uid);

      if (user != null) {
        await _populateCurrentUser(authResult.user);
        _status = ResultStatus.successful;
      } else {
        _status = ResultStatus.undefined;
      }
    } on FirebaseAuthException catch (e) {
      _status = ExceptionHandler.handleException(e);
    }
    return _status;
  }
```

Then, the viewmodel is this:
```
 Future signIn({
    @required String email,
    @required String password,
  }) async {
    setBusy(true);
    var result = await _authenticationService.signIn(
      email: email,
      password: password,
    );
    setBusy(false);
    if (result == ResultStatus.successful) {
      _navigationService.navigateTo(HomeRootScreen.route);
    } else {
      final errorMessage = ExceptionHandler.exceptionMessage(result);
      await _dialogService.showDialog(
        title: 'Sign up failed',
        description: errorMessage,
      );
    }
  }
```

My attempt:
```
 Future signIn({
    @required String email,
    @required String password,
  }) async {
    _firebaseAuth
        .signInWithEmailAndPassword(email: email, password: password)
        .then((authResult) {
      _populateCurrentUser(authResult.user);
      /*
      How to use the Response.success? I want to return the authResult.user
      if it's not equal to null
      */
      return authResult.user != null;
    }).catchError((e) =>
            Response.failure(type: ResponseType.failure, reason: e.code));
  }
```

My viewmodel:

```
  Future signIn({
    @required String email,
    @required String password,
  }) async {
    setBusy(true);
    _authenticationService
        .signIn(email: email, password: password)
        .then((result) => _navigationService.navigateTo(HomeRootScreen.route))
        .catchError((error) => _dialogService.showDialog(
              title: 'Sign up failed',
              description: error,
            ));
    setBusy(false);
  }
```

Reply:
I think your view model is right, as long as the dialog appears over the busy overlay. 

In your attempt, I expect the return you want is something like this:
```
        .then((authResult) {
      final result = Optional(authResult.user);
     // Probably should hold optional current user rather than null, too
     if (result.isPresent()) {
     return Response.success(true); // The statement `authResult.user != null returns a bool, right
    } else {
     return Response.failure(type: ResponseType.failure, reason: "No such account exists");
   }
})

Reply:
I will try this, thanks for all the help!